README.md
# Title: The Budget Hobo

The Budget Hobo allows the user to set up a budget, 
download your bank transactions, and view your budget performance.


# Demo
* Local Host

GitHub
* https://github.com/brandonhdez7/backendProject


# Table of Contents

* Description
* Technologies
* Challenges and Solutions
* MVP
* Our Stretch Goals
* Authors


# Description

## Program Features
The Budget Hobo allows the user to set up a budget, 
download your bank transactions, and view your budget performance.

## Profile
Click on the “Profile” link and the Basic Info
panel will be displayed. Enter your User Name, Email, and password twice.  
Click on Save to save or cancel to abort. 

<br>Click on Choose File to select a file from your local drive to add an image to your profile.  Click on Submit to save your image.

## Monthly Budget Set Up
Enter the amount you want to budget and click on "calculate" and your 
total Balance and Total Budget will auto-fill.  As expenses come in, enter or select your expense category. Enter your expense amount and click 
on add expense, your total expense will increase and your Total Balance will decrease.

## Dashboard
Click on the Dashboard to get a snapshot of your budget, transactions, and bank accounts. Graphs are available for a graphical view of your spending and budget categories.

## Bank Accounts
Click on connect to bank to use your bank credetials to log in to your bank.

# Technologies

* HTML - page layout and design
* CSS - page layout and design
* Bootstrap - page layout and design
* Fontawesome - page layout and design
* JavaScript - program functions
* JQuery - Budget and Graphs, API
* Node.js/Express - app generator, API
* Plaid - API, Secured Institution connection


# Challenges & Solutions
* Git & GitHub
* API
* Budget set up


# MVP
* Initial Layout
* Transaction and account balance API
* Quick Budget Set Up



# Stretch Goals
* Budget vs. Actual
* Increase database
* Income API


# Authors
* Brandon Hernandez
* Khanh Vu
* Brian Lafuente
* Ron Fletcher


history.txt
    1  ls
    2  cd development
    3  ls
    4  clr
    5  clear
    6  cd ..
    7  cd unit1
    8  ls
    9  open history.txt
   10  history
   11  open history.txt
   12  history
   13  open history.txt
   14  ls
   15  pwd
   16  cd 2018112
   17  ls
   18  cd unit1
   19  ls
   20  open history.txt
   21  ls
   22  cd etc
   23  cd ~
   24  pwd
   25  ls
   26  ls -la
   27  cd /
   28  ls
   29  ls -la
   30  ls
   31  cd applications
   32  ls
   33  cd ..
   34  clear
   35  cd etc
   36  pwd
   37  cd ..
   38  pwd
   39  ls
   40  cd etc
   41  ls
   42  cd apache
   43  cd apache2
   44  cd /users/
   45  ls
   46  pwd
   47  cd /etcapache2
   48  cd /etc/apache2
   49  clear
   50  cd ~/Delopmen/
   51  pwd
   52  ls -la
   53  cd git
   54  git add history.txt
   55  git status
   56  git commit -m "My first commit"
   57  cd ~
   58  cd finder
   59  ls
   60  cd Applications
   61  cd Utilities
   62  ls
   63  python -version
   64  python --version
   65  python python101.py
   66  ls
   67  cd..
   68  cd ..
   69  cd /
   70  ls
   71  cd Users
   72  ls
   73  cd /r
   74  cd ronfletcher1
   75  ls
   76  cd development
   77  ls
   78  cd 2018112
   79  ls
   80  cd unit1
   81  ls
   82  python python101.py
   83  ls
   84  git status
   85  git add python101.py
   86  git status
   87  git commit -m "First commit of the class for Ron Fletcher the python catcher"
   88  git status
   89  git remote add origin https://github.com/ronfletcher1/unit1immersive11-18-unit1.git\ngit push -u origin master
   90  history
   91  cd /
   92  pwd
   93  cd ~
   94  pwd
   95  ls 
   96  cd temp
   97  ls
   98  clear
   99  cd temp
  100  ls
  101  cd stuff
  102  ls
  103  cd things
  104  ls
  105  cd orange
  106  ls
  107  cd apple
  108  ls
  109  cd pear
  110  ls
  111  cd grape
  112  cd ../../../
  113  ls
  114  cd apple/pear/grape
  115  ls
  116  cd ..
  117  ls
  118  cd ../../../
  119  ls
  120  cd ../../
  121  ls
  122  ls -lr
  123  pwd
  124  ls
  125  cd stuff/things/orange/apple/pear/grape
  126  cd ..
  127  rmdir grape
  128  pwd
  129  cd ..
  130  rmdir pear
  131  cd..
  132  ls
  133  cd..
  134  cd ..
  135  rmdir apple
  136  cd..
  137  cd ..
  138  rmdir orange
  139  ls
  140  pwd
  141  cd ..
  142  cd ~
  143  ls
  144  pwd
  145  cd ..
  146  cd ~
  147  ls
  148  pwd
  149  python python101.py
  150  pwd
  151  ls
  152  cd development
  153  ls
  154  cd 2018112
  155  python python101.py
  156  cd unit1
  157  python python101.py
  158  python python101.pyp
  159  python1.1.py
  160  python python101.pyp
  161  python python101.py
  162  clear
  163  git add .
  164  git status
  165  git commit - m "Add operators"
  166  git commit - m "Add operator"
  167  git commit -m "Add operators"
  168  git push 
  169  ls
  170  ls -al
  171  rm -rf .git
  172  code history.txt
  173  code
  174  code history.txt
  175  pwd
  176  history
  177  git init
  178  git add .
  179  git status
  180  git commit -m "this is the new git commit exercise"
  181  git status
  182  python python101.py
  183  python python101.pyp
  184  python python101.py
  185  git status
  186  git log
  187  git remote add origin https://github.com/ronfletcher1/python-intro.git
  188  status
  189  git status
  190  git push -u origin master
  191  git add python101.py
  192  git status
  193  git commit -m "updated the python instructions"
  194  git status
  195  git push
  196  clear
  197  cd ~
  198  pwd
  199  ls
  200  mkdir CLI_basics
  201  ls
  202  cls
  203  cl
  204  pwd
  205  cl
  206  ls
  207  whoami
  208  clear
  209  python python101.py
  210  ls
  211  cd development
  212  ls
  213  cd 2018112
  214  ls
  215  cd unit1
  216  python python101.py
  217  clear
  218  python python101.py
  219  clear
  220  python python101.py
  221  clear
  222  python python101.py
  223  clear
  224  python python101.py
  225  clear
  226  python python101.py
  227  clear
  228  cd ~
  229  pwd
  230  mkdir CLI\ basics
  231  pwd
  232  cp -r CLI\ basics Development/
  233  pwd
  234  ls
  235  cd development
  236  pwd
  237  ls
  238  ~
  239  pwd
  240  ls
  241  cd development
  242  ls
  243  cd ..
  244  pwd
  245  cd Development
  246  pwd
  247  cd ..
  248  pwd
  249  rm -r CLI\ basics
  250  pwd
  251  ls
  252  cd Development
  253  pwd
  254  ls
  255  cd 2018112
  256  ls
  257  cd unit1
  258  mkdir CLI\ basics
  259  ls
  260  pwd
  261  git add CLI\ basics
  262  git status
  263  cler
  264  clear
  265  git status
  266  git commit -m "another practice"
  267  git status
  268  guess_the_number.py
  269  ls
  270  guess_the number.py
  271  python101.py
  272  ls
  273  cd ~
  274  ls
  275  cd ..
  276  ls
  277  cd users
  278  ls
  279  cd ~
  280  ls
  281  pwd
  282  cd..
  283  cd ..
  284  cd ../../
  285  pwd
  286  ls
  287  cd ..
  288  ls
  289  pwd
  290  cd Users
  291  cd ronfletcher1
  292  ls
  293  pwd
  294  cd downloads
  295  ls
  296  pwd
  297  ls
  298  cd ..
  299  ls
  300  cd development
  301  ls
  302  cd 2018112
  303  ls
  304  cd unit1
  305  ls
  306  forLoopsAndLists.py
  307  python forLoopsAndLists.py
  308  python guess_the number.py
  309  ls
  310  python guess_the_number.py
  311  python forLoopsAndLists.py
  312  pwd
  313  ls
  314  cd /
  315  pwd
  316  ls
  317  pwd
  318  ls
  319  cd users/ronfletcher1
  320  ls
  321  cd development
  322  ls
  323  cd 2018112
  324  cd cli basics
  325  cd CLI basics
  326  ls
  327  cd utils
  328  ls
  329  cd..
  330  cd .. 
  331  ls
  332  cd CLI\ basics
  333  ls
  334  cd ..
  335  pwd
  336  ls
  337  cd 2018112
  338  ls
  339  cd unit1
  340  ls
  341  clear
  342  python python stringPractice.py
  343  python stringPractice.py
  344  python python stringPractice.py
  345  ls
  346  python python stringPractice.py
  347  pwd
  348  ls
  349  history.txt
  350  python history.txt
  351  python notes.txt
  352  python python\ stringPractice.py
  353  cheese

budget.js
let y = []
let x = []
function removeExpense(arr) {
    var what, a = arguments, L = a.length, ax;
    while (L > 1 && arr.length) {
        what = a[--L];
        while ((ax= arr.indexOf(what)) !== -1) {
            arr.splice(ax, 1);
        }
    }
    return arr;
}

class UI {
    constructor() {
      this.budgetFeedback = document.querySelector(".budget-feedback");
      this.expenseFeedback = document.querySelector(".expense-feedback");
      this.budgetForm = document.getElementById("budget-form");
      this.budgetInput = document.getElementById("budget-input");
      this.budgetAmount = document.getElementById("budget-amount");
      this.expenseAmount = document.getElementById("expense-amount");
      this.balance = document.getElementById("balance");
      this.balanceAmount = document.getElementById("balance-amount");
      this.expenseForm = document.getElementById("expense-form");
      this.expenseInput = document.getElementById("expense-input");
      this.amountInput = document.getElementById("amount-input");
      this.expenseList = document.getElementById("expense-list");
      this.itemList = [];                 
      this.itemID = 0;
      this.submitBudgetForm = this.submitBudgetForm.bind(this)
    }
    submitBudgetForm(){
      // console.log("hello")
      const value = this.budgetInput.value;
      if(value==="" || value <0){
        this.budgetFeedback.classList.add('showItem')
        this.budgetFeedback.innerHTML = `<p> value cannot be empty or negative</p>`;
        const self = this;
        console.log(this)
        setTimeout(function(){
          console.log(this)
          console.log(self)
          self.budgetFeedback.classList.remove('showItem')
        }, 4000)
      }
      else{
        this.budgetAmount.textContent = value;
        this.budgetInput.value = '';
        this.showBalance()
      }
    }
    showBalance(){
      console.log("heyy")
      const expense = this.totalExpense();
      const total = parseInt(this.budgetAmount.textContent) - expense;
      this.balanceAmount.textContent = total;
      if(total < 0){
        this.balance.classList.remove('showGreen', 'showBlack');
        this.balance.classList.add('showRed')
      }    
      else if(total > 0){
        this.balance.classList.remove('showRed', 'showBlack');
        this.balance.classList.add('showGreen')
      }
      else if(total == 0){
        this.balance.classList.remove('showRed', 'showGreen');
        this.balance.classList.add('showBlack')
      }
    }
    submitExpenseForm(){
      const expenseValue = this.expenseInput.value;
      const amountValue = this.amountInput.value;
      if (expenseValue === '' || amountValue === '' || amountValue < 0){
        this.expenseFeedback.classList.add('showItem')
        this.expenseFeedback.innerHTML = `<p> value cannot be empty or negative</p>`;
        const self = this;
        setTimeout(function(){
          self.expenseFeedback.classList.remove('showItem')
        }, 4000)
      }
      else{
        let amount = parseInt(amountValue);
        this.expenseInput.value = '';
        this.amountInput.value = '';
  
        let expense = {
          id:this.itemID,
         title:expenseValue,
          amount:amount,
        }
        this.itemID++;
        this.itemList.push(expense)
        this.addExpense(expense)
        this.showBalance()
      }
    }
    
  addExpense(expense){
      console.log(expense)
    const div = document.createElement('div');
    div.classList.add('expense');
    div.innerHTML = `<div class="expense-item ">
  
    <h6 class="expense-title mb-0 text-uppercase list-item">${expense.title}</h6>
    <h5 class="expense-amount mb-0 list-item">${expense.amount}</h5>
  
    <div class="expense-icons list-item">
  
     <a href="#" class="edit-icon mx-2" data-id="${expense.id}">
      <i class="fas fa-edit"></i>
     </a>
     <a href="#" class="delete-icon" data-id="${expense.id}">
      <i class="fas fa-trash"></i>
     </a>
    </div>
   </div>`;
   this.expenseList.appendChild(div)
   
   x.push(expense.title)
   y.push(expense.amount)
   pieGraph(x, y)
}



totalExpense(){
    let total = 0;
    if(this.itemList.length>0){
    total = this.itemList.reduce(function(acc, curr){
        acc += curr.amount;
        return acc;
    },0)
    }
    this.expenseAmount.textContent = total;
    return total;
}
editExpense(element){
    let id = parseInt(element.dataset.id);
    let parent = element.parentElement.parentElement.parentElement;
    
    this.expenseList.removeChild(parent);
    // delete from the dom
    let expense = this.itemList.filter(function(item){
    return item.id === id;
    });
    this.expenseInput.value = expense[0].title;
    this.amountInput.value = expense[0].amount;
    // delete from the list
    let tempList = this.itemList.filter(function(item){
    return item.id !== id;
    })
    this.itemList = tempList;
    this.showBalance();
    removeExpense(x, expense[0].title)
    removeExpense(y, expense[0].amount)
    pieGraph(x, y)
}
deleteExpense(element){
    // console.log(JSON.parse(element))
    // console.log(element.title)
    let id = parseInt(element.dataset.id);
    let parent = element.parentElement.parentElement.parentElement;
    let expense = this.itemList.filter(function(item){
        return item.id === id;
        });
    console.log(expense)
    this.expenseList.removeChild(parent);

    let tempList = this.itemList.filter(function(item){
    return item.id != id;
    })
    // console.log(tempList[0].title)
    this.itemList = tempList;
    this.showBalance();
    removeExpense(x, expense[0].title);
    removeExpense(y, expense[0].amount);
    pieGraph(x, y)
}
}
function pieGraph(title, amount){

    var ctx = document.getElementById("myChart3").getContext('2d');
     var myChart = new Chart(ctx, {
         type: 'doughnut',
         data: {
             labels: title,
             datasets: [{ 
                 label: '# of Votes',
                 data: amount,
                 backgroundColor: [
                     'rgba(255, 99, 132, 0.2)',
                     'rgba(54, 162, 235, 0.2)',
                     'rgba(255, 206, 86, 0.2)',
                     'rgba(75, 192, 192, 0.2)',
                     'rgba(153, 102, 255, 0.2)',
                     'rgba(255, 159, 64, 0.2)'
                 ],
                 borderColor: [
                     'rgba(255,99,132,1)',
                     'rgba(54, 162, 235, 1)',
                     'rgba(255, 206, 86, 1)',
                     'rgba(75, 192, 192, 1)',
                     'rgba(153, 102, 255, 1)',
                     'rgba(255, 159, 64, 1)'
                 ],
                 borderWidth: 1
             }]
         },
         options: {
             legend:{
                 position: 'right',
                 scales: {
                     yAxes: [{
                         ticks: {
                             beginAtZero:true
                         }
                     }]
                 }
             
             }   
         }
     })
}

function EventListener(){
const budgetForm = document.getElementById('budget-form')
const expenseForm = document.getElementById('expense-form')
const expenseList = document.getElementById('expense-list')

const ui = new UI()

budgetForm.addEventListener('submit', function(event){
    console.log("loooo")
    event.preventDefault();
    const totalBudget = document.getElementById('budget-input').value;
    const expense = document.querySelector('#expense-input').value;
    console.log(expense)
    const amount = document.querySelector('#amount-input').value;
    const name = document.querySelector('#name-field').value;
    const ajaxRequest = $.ajax({
        type: 'POST',
        url: 'http://localhost:4000/formBudget',
        data: {
            totalBudget,
            expense,
            amount,
            name
        },
        dataType: 'json',
    });
    ajaxRequest.then((data)=>{
        console.log(data)
        // if(data.msg === "success"){
            ui.submitBudgetForm();
        // }else{
        //     throw "ajax failed :(";
        // }
        
    })
})

expenseForm.addEventListener('submit', function(event){
    event.preventDefault();
    ui.submitExpenseForm();
})

expenseList.addEventListener('click', function(event){
    if(event.target.parentElement.classList.contains('edit-icon')){
    ui.editExpense(event.target.parentElement);
    }
    else if(event.target.parentElement.classList.contains('delete-icon')){
    ui.deleteExpense(event.target.parentElement);
    }
})
}



document.addEventListener('DOMContentLoaded', function(){
    EventListener();
})






// upload image

// $("#profileImage1").click(function(e) {
//     $("#imageUpload").click();
// });

// function fasterPreview( uploader ) {
//     if ( uploader.files && uploader.files[0] ){
//           $('#profileImage1').attr('src', 
//              window.URL.createObjectURL(uploader.files[0]) );
//     }
// }

// $("#imageUpload").change(function(){
//     fasterPreview( this );
// });

// const aPromise = new Promise((resolve, reject)=>{
//     setTimeout(()=>{
//         resolve("100 ms have passed")
//     },100)
// })


